app/src/main/java/com/github/geohunt/app/data/repository/ClaimRepository.kt

```diff
@@ -0,0 +1,159 @@
+import com.github.geohunt.app.data.exceptions.auth.UserNotLoggedInException
+import com.github.geohunt.app.model.database.api.Location
+import com.github.geohunt.app.data.local.LocalPicture
+import com.github.geohunt.app.data.network.firebase.models.FirebaseClaim
+import com.github.geohunt.app.model.Challenge
+import com.github.geohunt.app.model.Claim
+import com.github.geohunt.app.model.User
+import com.github.geohunt.app.model.points.GaussianPointCalculator
+import com.github.geohunt.app.model.points.PointCalculator
+import com.github.geohunt.app.utility.DateUtils
+import com.google.firebase.database.FirebaseDatabase
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.tasks.asDeferred
+import kotlinx.coroutines.tasks.await
+
+/**
+ * Contains methods for claim/unclaim a challenge, as well as retrieving the list
+ * of claim for a particular challenge or a particular user
+ */
+class ClaimRepository(
+    private val authRepository: AuthRepository,
+    private val imageRepository: ImageRepository,
+    private val database: FirebaseDatabase = FirebaseDatabase.getInstance(),
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
+    private val pointCalculatorMap: Map<Challenge.Difficulty, PointCalculator> = mapOf(
+        Challenge.Difficulty.EASY to GaussianPointCalculator(0.20),
+        Challenge.Difficulty.MEDIUM to GaussianPointCalculator(0.15),
+        Challenge.Difficulty.EASY to GaussianPointCalculator(0.10)
+    ).withDefault { GaussianPointCalculator(0.10) }
+) {
+    private fun getChallengeRefFromId(id: String) {
+        database.getReference("claims/$id")
+    }
+
+    private fun FirebaseClaim.asExternalModel(id: String): Claim = Claim(
+        id = id,
+        parentChallengeId = cid!!,
+        claimerId = uid!!,
+        claimDate = DateUtils.localFromUtcIso8601(time!!),
+        distance = distance,
+        awardedPoints = awardedPoints,
+        photoUrl = photoUrl!!
+    )
+
+    /**
+     * Retrieve a list of all claims id for a specific user, useful when lazy loading
+     */
+    suspend fun getClaimsIdByUser(user: User) : List<String> =
+        database.getReference("claimsByUser/${user.id}")
+            .get()
+            .await()
+            .run {
+                children.mapNotNull { it.getValue(String::class.java) }
+            }
+
+    suspend fun doesClaims(challenge: Challenge) : Boolean {
+        authRepository.requireLoggedIn()
+        val currentUser = authRepository.getCurrentUser()
+
+        return getClaimsByUser(currentUser)
+            .any { it.parentChallengeId == challenge.id }
+    }
+
+    /**
+     * Retrieve the score for a given user
+     */
+    suspend fun getScoreFromUser(user: User) : Long =
+        getClaimsByUser(user).fold(0L) { acc, claim -> acc + claim.awardedPoints }

```

Comment:
What about using `.sumOf { it.awardedPoints }`?

Reply:
It is not missleading enough 🙃 No jk nice catch